presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedQueriesWithExchangeMaterialization.java

```diff
@@ -188,6 +188,99 @@ public void testCorrelatedScalarSubqueriesWithScalarAggregation()
         // Anonymous row type is not supported in Hive
     }
 
+    @Test
+    public void testExchangeMaterializationWithConstantFolding()
+    {
+        try {
+            assertUpdate(
+                    // bucket count has to be different from materialized bucket number
+                    "CREATE TABLE test_constant_folding_lineitem_bucketed\n" +
+                            "WITH (bucket_count = 17, bucketed_by = ARRAY['partkey_mod_9', 'partkey', 'suppkey', 'suppkey_varchar']) AS\n" +
+                            "SELECT partkey % 9 partkey_mod_9, partkey, suppkey, CAST(suppkey AS VARCHAR) suppkey_varchar, comment FROM lineitem",
+                    "SELECT count(*) from lineitem");
+            assertUpdate(
+                    "CREATE TABLE test_constant_folding_partsupp_unbucketed AS\n" +
+                            "SELECT partkey % 9 partkey_mod_9, partkey, suppkey, CAST(suppkey AS VARCHAR) suppkey_varchar, comment FROM partsupp",
+                    "SELECT count(*) from partsupp");
+
+            // one constant, third position (suppkey BIGINT)
+            assertQuery(
+                    getSession(),
+                    "SELECT lineitem.partkey, lineitem.suppkey, lineitem.comment lineitem_comment, partsupp.comment partsupp_comment\n" +
+                            "FROM test_constant_folding_lineitem_bucketed lineitem JOIN test_constant_folding_partsupp_unbucketed partsupp\n" +
+                            "ON\n" +
+                            "  lineitem.partkey = partsupp.partkey AND\n" +
+                            "  lineitem.partkey_mod_9 = partsupp.partkey_mod_9 AND\n" +
+                            "  lineitem.suppkey = partsupp.suppkey AND\n" +
+                            "  lineitem.suppkey_varchar = partsupp.suppkey_varchar\n" +
+                            "WHERE lineitem.suppkey = 42",
```

Comment:
Could you please double check if this causes a partitioning constant folding? The `lineitem.suppkey` field is referenced in the final projection. So it looks like this channel is never removed through the query flow. And if the channel is there - we prefere to use channel in partitioning than a constant (see https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java#L316)